hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/associations/TokenAssociationsDumpUtils.java

```diff
@@ -0,0 +1,149 @@
+                    associations, key -> TokenAssociationId.fromMod(key.getKey()), TokenAssociation::fromMod);
+            reportOnTokenAssociations(writer, dumpableTokenRelations);
+            System.out.printf(
+                    "=== mod token associations report is %d bytes at checkpoint %s%n",
+                    writer.getSize(), checkpoint.name());
+        }
+    }
+
+    public static void dumpMonoTokenRelations(
+            @NonNull final Path path,
+            @NonNull final VirtualMap<OnDiskKey<OnDiskTokenRel>, OnDiskTokenRel> associations,
+            @NonNull final DumpCheckpoint checkpoint) {
+        try (@NonNull final var writer = new Writer(path)) {
+            final var dumpableTokenRelations =
+                    gatherTokenRelations(associations, TokenAssociationId::fromMono, TokenAssociation::fromMono);
+            reportOnTokenAssociations(writer, dumpableTokenRelations);
+            System.out.printf(
+                    "=== mono token associations report is %d bytes at checkpoint %s%n",
+                    writer.getSize(), checkpoint.name());
+        }
+    }
+
+    @NonNull
+    private static <K extends VirtualKey, V extends VirtualValue>
+            Map<TokenAssociationId, TokenAssociation> gatherTokenRelations(
+                    @NonNull final VirtualMap<K, V> source,
+                    @NonNull final Function<K, TokenAssociationId> keyMapper,
+                    @NonNull final Function<V, TokenAssociation> valueMapper) {
+        final var r = new HashMap<TokenAssociationId, TokenAssociation>();
+        final var threadCount = 8;
+        final var tokenAssociations = new ConcurrentLinkedQueue<Pair<TokenAssociationId, TokenAssociation>>();
+        try {
+            VirtualMapLike.from(source)
+                    .extractVirtualMapData(
+                            getStaticThreadManager(),
+                            p -> tokenAssociations.add(
+                                    Pair.of(keyMapper.apply(p.left()), valueMapper.apply(p.right()))),
+                            threadCount);
+        } catch (final InterruptedException ex) {
+            System.err.println("*** Traversal of token associations virtual map interrupted!");
+            Thread.currentThread().interrupt();
+        }
+        tokenAssociations.forEach(
+                tokenAssociationPair -> r.put(tokenAssociationPair.key(), tokenAssociationPair.value()));
+        return r;
+    }
+
+    private static void reportOnTokenAssociations(
+            @NonNull final Writer writer, @NonNull final Map<TokenAssociationId, TokenAssociation> associations) {
+        writer.writeln(formatHeader());
+        associations.entrySet().stream()
+                .sorted(Map.Entry.comparingByKey())
+                .forEach(e -> formatTokenAssociation(writer, e.getValue()));
+        writer.writeln("");
+    }
+
+    @NonNull
+    private static String formatHeader() {
+        return "accountId,tokenId,"
```

Comment:
Sorry for the late comment. Shouldn't we delete the `accountId,tokenId,` part? It seems that this will generate duplicate columns 🤔

Reply:
At first I was thinking about the same.. @david-bakin-sl thoughts on that?

The other option could be to remove `accountId` and `tokenId` from the `fieldFormatters` [here](https://github.com/hashgraph/hedera-services/blob/10918-token-associations-modular-dumper/hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/associations/TokenAssociationsDumpUtils.java#L118-L119)

Reply:
Good catch.  I think just remove the `"accountId,tokenId,"+` here.

I knew it was going to be a problem asking you to write code with no way to test it.  But good code review counts, thanks Valentin!  (I should test it - but I can't run the rounds for technical reasons at this time.  I'll see if I can ask someone to test this - but the people who could do it are pretty busy right now on testing event rounds. 
 Even so don't hold up the approvals+merge for it.)